driver-core/src/main/com/mongodb/internal/connection/OperationContext.java

```diff
@@ -23,12 +34,87 @@
 public class OperationContext {
     private static final AtomicLong NEXT_ID = new AtomicLong(0);
     private final long id;
+    private final ServerDeprioritization serverDeprioritization;
 
     public OperationContext() {
         id = NEXT_ID.incrementAndGet();
+        serverDeprioritization = new ServerDeprioritization();
     }
 
     public long getId() {
         return id;
     }
+
+    /**
+     * @return The same {@link ServerDeprioritization} if called on the same {@link OperationContext}.
+     */
+    public ServerDeprioritization getServerDeprioritization() {
+        return serverDeprioritization;
+    }
+
+    public static final class ServerDeprioritization {
+        @Nullable
+        private ServerAddress candidate;
+        private final Set<ServerAddress> deprioritized;
+
+        private ServerDeprioritization() {
+            candidate = null;
+            deprioritized = new HashSet<>();
+        }
+
+        ServerSelector apply(final ServerSelector selector) {
+            return new DeprioritizingSelector(selector);
+        }
+
+        void updateCandidate(final ServerAddress serverAddress, final ClusterType clusterType) {
+            candidate = isEnabled(clusterType) ? serverAddress : null;
+        }
+
+        public void onAttemptFailure(final Throwable failure) {
+            if (candidate != null && !(failure instanceof MongoConnectionPoolClearedException)) {
+                deprioritized.add(candidate);
+            }
+            candidate = null;
+        }
+
+        private static boolean isEnabled(final ClusterType clusterType) {
+            return clusterType == ClusterType.SHARDED;
+        }
+
+        /**
+         * {@link ServerSelector} requires thread safety, but that is only because a user may specify
+         * {@link com.mongodb.connection.ClusterSettings.Builder#serverSelector(ServerSelector)},
+         * which indeed may be used concurrently. {@link DeprioritizingSelector} does not need to be thread-safe.
+         */
+        private final class DeprioritizingSelector implements ServerSelector {
+            private final ServerSelector wrapped;
+
+            private DeprioritizingSelector(final ServerSelector wrapped) {
+                this.wrapped = wrapped;
+            }
+
+            @Override
+            public List<ServerDescription> select(final ClusterDescription clusterDescription) {
+                if (isEnabled(clusterDescription.getType())) {
```

Comment:
Is this check needed here for correctness, and is there a possible noticeable effect on performance?

This check would be clearer if inverted:
```
if (!isEnabled(clusterDescription.getType())) {
    return wrapped.select(clusterDescription);
}
// followed by the non-indented deprioritization logic
```